modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvPartitionStorageGcTest.java

```diff
@@ -0,0 +1,132 @@
+        assertNull(storage.pollForVacuum(clock.now()));
+    }
+
+    @Test
+    void testSingleValueStorage() {
+        addAndCommit(BINARY_ROW);
+
+        assertNull(storage.pollForVacuum(clock.now()));
+    }
+
+    @Test
+    void testRegularPoll() {
+        HybridTimestamp firstCommitTs = addAndCommit(BINARY_ROW);
+
+        HybridTimestamp tsBetweenCommits = clock.now();
+
+        HybridTimestamp secondCommitTs = addAndCommit(BINARY_ROW);
+
+        // Data is still visible for older timestamps.
+        assertNull(storage.pollForVacuum(firstCommitTs));
+
+        assertNull(storage.pollForVacuum(tsBetweenCommits));
+
+        // Once a low watermark value becomes equal to second commit timestamp, previous value
+        // becomes completely inaccessible and should be purged.
+        BinaryRowWithRowId row = storage.pollForVacuum(secondCommitTs);
+
+        assertNotNull(row);
+
+        assertRowMatches(row.binaryRow(), BINARY_ROW);
+
+        // Read from the old timestamp should return null.
+        assertNull(read(ROW_ID, firstCommitTs));
+
+        // Read from the newer timestamp should return last value.
+        assertRowMatches(read(ROW_ID, secondCommitTs), BINARY_ROW);
+    }
+
+    @Test
+    void testPollFromUnderTombstone() {
+        addAndCommit(BINARY_ROW);
+        HybridTimestamp secondCommitTs = addAndCommit(null);
+
+        BinaryRowWithRowId row = storage.pollForVacuum(secondCommitTs);
+
+        assertNotNull(row);
+        assertRowMatches(row.binaryRow(), BINARY_ROW);
+
+        assertNull(read(ROW_ID, secondCommitTs));
+
+        // Check that tombstone is also deleted from the partition. It must be empty at this point.
+        assertNull(storage.closestRowId(ROW_ID));
+    }
+
+    @Test
+    void testDoubleTombstone() {
+        addAndCommit(BINARY_ROW);
+        addAndCommit(null);
+        HybridTimestamp lastCommitTs = addAndCommit(null);
+
+        BinaryRowWithRowId row = storage.pollForVacuum(lastCommitTs);
+
+        assertNotNull(row);
+        assertRowMatches(row.binaryRow(), BINARY_ROW);
+
+        assertNull(read(ROW_ID, lastCommitTs));
+
+        // Check that all tombstones are deleted from the partition. It must be empty at this point.
+        assertNull(storage.closestRowId(ROW_ID));
+    }
+
+    @Test
+    void testManyOldVersions() {
+        addAndCommit(BINARY_ROW);
+
+        BinaryRow binaryRow2 = binaryRow(KEY, new TestValue(50, "50"));
+
+        addAndCommit(binaryRow2);
+
+        HybridTimestamp lowWatermark = addAndCommit(null);
+
+        // Poll the oldest row.
+        BinaryRowWithRowId row = pollForVacuum(lowWatermark);
+
+        assertNotNull(row);
+        assertRowMatches(row.binaryRow(), BINARY_ROW);
+
+        // Poll the next oldest row.
+        row = pollForVacuum(lowWatermark);
+
+        assertNotNull(row);
+        assertRowMatches(row.binaryRow(), binaryRow2);
+
+        // Nothing else to poll.
+        assertNull(pollForVacuum(lowWatermark));
+    }
+}
```

Comment:
Let's add a test (or a few) that models the real threading model that will be used in production code: one thread modifies the storage, another one acts as a GC. This might help catch visibility bugs (if, for some reason, `pollForVacuum()` never sees changes made by other methods).

Reply:
I should add more scenarios to AbstractMvPartitionStorageConcurrencyTest, is this what you mean? Yeah, I guess so, let's be paranoid, I like that

Reply:
Scenarios that are closer to the real usages will be tested later, when we implement the component that actually calls this new method.

Reply:
Sure, I'm ok with `AbstractMvPartitionStorageConcurrencyTest`